lib/entry/crossDCsStreamPersistence.js

```diff
@@ -0,0 +1,218 @@
+/**
+ * Created by lilach.maliniak on 18/09/2016.
+ */
+var persistenceFormat = require('./../../common/PersistenceFormat');
+var qio = require('q-io/fs');
+//var sessionManager = require('./../SessionManager');
+var logger = require('../../common/logger').getLogger("CrossDCsStreamPersistence");
+var config = require('./../../common/Configuration');
+var ErrorUtils = require('./../utils/error-utils');
+var _ = require('underscore');
+var Q = require('q');
+var glob = require('glob');
+var path = require('path');
+var backendClient = require('../BackendClientFactory.js').getBackendClient();
+var PlaylistGenerator = require('./../playlistGenerator/PlaylistGenerator');
+const persistenceFileName = config.get('persistenceFileName');
+const persistenceCheckIntervalMillisec = 1000 * config.get('persistenceCheckIntervalSec');
+const kalturaTypes = require('./../kaltura-client-lib/KalturaTypes');
+const presentationEndTimeRegex = /("presentationEndTime":\s*)(\d+)/;
+
+// todo: add sorted list according to time left for removal of presentationEndTime and expirationTime properties
+// Add processEndedEntries or use processStoppedEntries to handle the removal of above for DVR support.
+// Should the future time set to presentationEndTime be derived from real liveWindowDuration value?
+// todo: remove the presentationEndTime and expirationTime from playlist.json in PlaylistGenerat
+
+function playlistPath(entryId) {
+    return path.join(persistenceFormat.getEntryBasePath(entryId), persistenceFormat.getMasterManifestName());
+}
+
+class CrossDCsStreamPersistence {
+
+    constructor() {
+        this._entriesToProcess = {};
+        this._lastProcessedTimestamp = Date.now();
+    }
+
+    streamPersistenceCheck(entryId) {
+        logger.debug(`[${entryId}] starting stream persistence check`);
+
+        // todo: consult with Guy whether to filter received nodes objects according to status...
+        //entryServerNodeFilter.statusEqual = KalturaLiveStatus.PLAYABLE or KalturaLiveStatus.BROADCASTING;
+        return backendClient.getLiveEntryServerNodes(entryId)
+            .then((serverObjs) => {
+                return _.reduce(serverObjs, (memo, dc) => {
+                    return (dc.status === kalturaTypes.KalturaEntryServerNodeStatus.PLAYABLE) || memo;
+                }, false);
+            })
+            .then((isLive) => {
+                return this.playlistUpdate(entryId, isLive);
+            })
+            .catch((err) => {
+                logger.error(`[${entryId}] persistence check failed. Error: [${ErrorUtils.error2string(err)}]`);
+                return Q.reject(err);
+            });
+
+    }
+
+    playlistUpdate(entryId, isLive) {
+
+        let now = Date.now();
+        let path = playlistPath.call(this, entryId);
+        let ended = !isLive;
+
+        return qio.read(path)
+            .then((content) => {
+                let playlist = content;
+
+                if (ended) {
```

Comment:
i think it's better to do if (!live) {
   logger.debug(`[${entryId}] ENDED`);
   playlist = playlist.replace(presentationEndTimeRegex, `$1 ${now}`);
  return qio.write(path, playlist).then( ()=> {
                   return this.removeEntry(entryId, true);
  }
}